Initial modifications for using only KBase auth service in the

kkellerlbl opened the pull request:
[WIP] DO NOT MERGE

...in the auth client.

This first pass only looks at getting a token given
a userid/password (and probably breaks ssh key-based login).

kkellerlbl committed:
Initial modifications for using only KBase auth service in the
auth client.  This first pass only looks at getting a token given
a userid/password (and probably breaks ssh key-based login).

MrCreosote commented:
> and probably breaks ssh key-based login

Don't care, AFAIK that's totally unused. It's not supported in any of the other clients.

MrCreosote commented on Bio-KBase-Auth/lib/Bio/KBase/AuthToken.pm:
```diff
@@ -475,8 +479,9 @@ sub validate {
+        warn $Bio::KBase::Auth::AuthSvcHost;
 	unless ( $vars{'SigningSubject'} =~ /^\Q$Bio::KBase::Auth::AuthSvcHost\E/) {
-	    die "Token signed by unrecognized source: ".$vars{'SigningSubject'};
+	    warn "Token signed by unrecognized source: ".$vars{'SigningSubject'};
```
This should just go - knowing the SS implies parsing the token.

kkellerlbl replied:
That's what I figured, I'll remove it in the culling phase.

MrCreosote commented on Bio-KBase-Auth/lib/Bio/KBase/AuthToken.pm:
```diff
@@ -585,6 +592,54 @@ sub go_request {
+
+    my $json;
+    eval {
+	my $baseurl = $Bio::KBase::Auth::AuthSvcHost;
```
The url should be passed in to the new() method rather than from the config. I assume you know, but just checking.

kkellerlbl replied:
Yes, I know.  I left it for now just so I could do some of my own testing.

I did find out today that there actually is a way to put this parameter in a config file: it comes from the .kbase_config file.  We can leave it in there or I can move it to the deployment.cfg file, I'm not sure it matters.  But I'm not happy with how it currently gets configured, it's a little opaque, so I do want to modify that to be clearer.

MrCreosote replied:
In the other clients, the url is passed in, so my feeling is that we should do the same here. The servers should look for auth-server-url in their section of the deploy.cfg as we discussed. 

We might need to alter kb-login to take an auth url temporarily though - not sure. It might be enough just to provide instructions on how to modify the script short term, since the script will stop working once we switch to the new auth server.

MrCreosote replied:
On the other hand, maybe it is simpler in the short term to grab the url from the .kbase_config (or .authrc, I forget the name. We should use the same name as the py client (which is embedded in baseclient.py).

MrCreosote commented on Bio-KBase-Auth/lib/Bio/KBase/AuthToken.pm:
```diff
@@ -585,6 +592,54 @@ sub go_request {
+#	my $client = LWP::UserAgent->new(default_headers => $headers);
+	my $client = LWP::UserAgent->new();
+	$client->timeout(5);
+	$client->ssl_opts(verify_hostname => 0);
```
This seems like a bad idea - why aren't we verifying the host?

kkellerlbl replied:
Dunno.  This was just copy/paste from the go_request method.  It could be an artifact from when we didn't have the wildcard cert on ci?  Anyway, it probably makes sense to remove this option.

MrCreosote commented on Bio-KBase-Auth/lib/Bio/KBase/AuthToken.pm:
```diff
@@ -585,6 +592,54 @@ sub go_request {
+	    die $response->status_line;
+	}
+	$json = decode_json( $response->content());
+	$json = $self->_SquashJSONBool( $json);
```
I doubt this is necessary

kkellerlbl replied:
Probably not.  I want to read more on what SquashBool does before removing it though.

MrCreosote replied:
It just maps JSON:Bool (or whatever the name is) types to 0 or 1.

MrCreosote commented on Bio-KBase-Auth/lib/Bio/KBase/AuthToken.pm:
```diff
@@ -585,6 +592,54 @@ sub go_request {
+#	my %headers;
+#        warn $p{'path'};
+	unless ($p{'path'}) {
+	    die "No path specified";
```
I'm confused re baseurl vs path. Seems like it'd be simpler to just pass in the entire auth url like the py client does it.

kkellerlbl replied:
This code is a huge mess.  For my first pass I didn't clean it up, but it's clear that it needs work to make more readable.  I'll get to that when I fix up how the config file gets read.

MrCreosote commented:
Are there tests for this, and if so do they still pass (I assume that setup will need to change...)

kkellerlbl commented:
Thanks Gavin!  I'll reply to the line responses individually.
- I figured we didn't need to support ssh key auth.  I'll probably just remove all of that code later.
- I had assumed there were no tests but I admit I didn't actually look.  I'll look tomorrow.

kkellerlbl commented:
It seems like you haven't found any huge red flags so far (at least none I didn't already know about), so that's good!

kkellerlbl commented:
There are tests!  I didn't see them initially because they're in the Bio-KBase-Auth/t dir instead of in client-tests at the top level.  I'll take a look at them tomorrow (and also see if there are tests for the Python client that the Perl tests aren't covering and add them).

kkellerlbl committed:
validate() uses server

The validate() method now uses the specified auth server instead of
validating locally.  Starting to remove code for using RSA keys.

kkellerlbl committed:
More cleanup

Remove remnants of ssh key auth
Starting to look at configuration options

kkellerlbl commented on Bio-KBase-Auth/lib/Bio/KBase/AuthToken.pm:
```diff
@@ -140,8 +109,7 @@ sub new {
 		$self->token( $c{'authentication.token'});
-# the client no longer validates tokens
-#		$self->validate();
+		$self->validate();
```
Not sure we need this here

kkellerlbl committed:
Bumped version
Some more cleanup
Don't attempt to validate token if no token was provided

kkellerlbl committed:
Get original tests to pass

Bug fix in AuthToken::validate (need to return undef and set error_message)
Remove RSA key tests, deprecate AuthUser tests

kkellerlbl committed:
Add config file to tests

Beginning to add a config file to tests (so that credentials do
not need to be stored in the repo)
Added skeleton test.cfg file

kkellerlbl committed:
Changes for testing and configuration

Removing AuthConstants, AuthUser libraries
Removed auth credentials from test scripts, moved to test.cfg (added with example values)
Added auth_svc parameter to AuthToken::new(), overrides any default values

MrCreosote commented:
Note that as part of testing we should make sure kb-login and kb-whoami work

MrCreosote commented on Bio-KBase-Auth/lib/Bio/KBase/AuthToken.pm:
```diff
@@ -442,145 +272,92 @@ sub canonical_time {
+            my $json = decode_json( $response->content());
+#	    $json = $self->_SquashJSONBool($json);
+
+            # write the sha1 hash of the token into the cache
```
I'm using sha256 in python and java

MrCreosote commented on Bio-KBase-Auth/t/testAuthToken.t:
```diff
@@ -89,188 +32,102 @@ my %old_config = map { $_ =~ s/authentication\.//; $_ => $Bio::KBase::Auth::Conf
-    %new = %old_config;
-    foreach $key ( 'user_id','password','keyfile','keyfile_passphrase','client_secret','token') {
+    my %new = %old_config;
+    foreach my $key ( 'user_id','password','keyfile','keyfile_passphrase','client_secret','token') {
```
a comment as requested by keith
